kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/ImmutableCollection.kt

```diff
@@ -16,24 +16,96 @@
+     *
+     * @return a new persistent collection with elements of the specified [elements] collection added;
+     * or this instance if no modifications were made in the result of this operation.
+     */
     fun addAll(elements: Collection<@UnsafeVariance E>): PersistentCollection<E>
 
+    /**
+     * Returns the result of removing a single appearance of the specified [element] from this collection.
+     *
+     * @return a new persistent collection with a single appearance of the specified [element] removed;
+     * or this instance if there is no such element in this collection.
+     */
     fun remove(element: @UnsafeVariance E): PersistentCollection<E>
 
+    /**
+     * Returns the result of removing all elements in this collection that are also
+     * contained in the specified [elements] collection.
+     *
+     * @return a new persistent collection with elements in this collection that are also
+     * contained in the specified [elements] collection removed;
+     * or this instance if no modifications were made in the result of this operation.
+     */
     fun removeAll(elements: Collection<@UnsafeVariance E>): PersistentCollection<E>
 
+    /**
+     * Returns the result of removing all elements in this collection that match the specified [predicate].
+     *
+     * @return a new persistent collection with elements matching the specified [predicate] removed;
+     * or this instance if no elements match the predicate.
+     */
     fun removeAll(predicate: (E) -> Boolean): PersistentCollection<E>
 
+    /**
+     * Returns an empty persistent collection.
+     */
     fun clear(): PersistentCollection<E>
 
+    /**
+     * A generic builder of the persistent collection. Builder exposes its modification operations through the [MutableCollection] interface.
+     *
+     * Builders are reusable, that is [build] method can be called multiple times with modifications between these calls.
+     * However, modifications applied do not affect previously built persistent collection instances.
+     */
     interface Builder<E>: MutableCollection<E> {
+        /**
+         * Returns a persistent collection with the same contents as this builder.
+         *
+         * This method can be called multiple times.
+         * If operations applied on this builder have caused no modifications relative to the previously returned persistent collection,
+         * it returns the same previously returned persistent collection instance.
+         * In case this method is called for the first time on this builder and all operations applied have caused no modifications,
+         * it returns the same persistent collection instance this builder was obtained from.
+         */
         fun build(): PersistentCollection<E>
     }
 
+    /**
+     * Returns a new builder with the same contents as this collection.
+     *
+     * The builder can be used to efficiently perform multiple modification operations.
```

Comment:
We can expand, why it is efficient here:
- it doesn't copy the collection data structures when being created. 
- it creates new data structures only when modifications are made the first time, and reuses these structures for subsequent modifications until [build] is called.

Reply:
done